eacl/target.go

```diff
@@ -1,200 +1,174 @@
+//
+// See also [NewTarget], [NewTargetWithKeys].
+func NewTargetWithKey(key neofscrypto.PublicKey) Target {
+	return NewTargetWithKeys([]neofscrypto.PublicKey{key})
 }
 
-// Marshal marshals Target into a protobuf binary form.
-func (t *Target) Marshal() ([]byte, error) {
-	return t.ToV2().StableMarshal(nil), nil
+// NewTargetWithKeys returns Target for the given list of public keys.
+//
+// See also [NewTarget], [NewTargetWithKey].
+func NewTargetWithKeys(publicKeys []neofscrypto.PublicKey) Target {
+	return NewTarget(nil, publicKeys)
 }
 
-// Unmarshal unmarshals protobuf binary representation of Target.
-func (t *Target) Unmarshal(data []byte) error {
-	fV2 := new(v2acl.Target)
-	if err := fV2.Unmarshal(data); err != nil {
-		return err
+// readFromV2 reads Target from the [v2acl.Target] messages. Returns an error if
+// any message is malformed according to the NeoFS API V2 protocol. Behavior is
+// forward-compatible:
+//   - unknown enum values are considered valid
+//   - unknown format of binary public keys is considered valid
+//
+// The argument MUST be non-empty.
+func (t *Target) readFromV2(ms []v2acl.Target) error {
+	if len(ms) == 0 {
+		panic("empty slice of targets")
 	}
 
-	*t = *NewTargetFromV2(fV2)
+	for i := range ms {
+		role := ms[i].GetRole()
+		keys := ms[i].GetKeys()
 
-	return nil
-}
+		if role == 0 && len(keys) == 0 {
+			return fmt.Errorf("invalid target #%d: neither role nor public keys are set", i)
+		}
 
-// MarshalJSON encodes Target to protobuf JSON format.
-func (t *Target) MarshalJSON() ([]byte, error) {
-	return t.ToV2().MarshalJSON()
-}
+		if role != 0 && len(keys) != 0 {
+			return fmt.Errorf("invalid target #%d: both role and public keys are set", i)
+		}
 
-// UnmarshalJSON decodes Target from protobuf JSON format.
-func (t *Target) UnmarshalJSON(data []byte) error {
-	tV2 := new(v2acl.Target)
-	if err := tV2.UnmarshalJSON(data); err != nil {
-		return err
-	}
+		for j := range keys {
+			if len(keys[j]) == 0 {
+				return fmt.Errorf("invalid target #%d: empty key #%d", i, j)
+			}
+		}
 
-	*t = *NewTargetFromV2(tV2)
+		if role != 0 {
+			t.roles = append(t.roles, Role(role))
+		} else {
+			t.keys = append(t.keys, keys...)
+		}
+	}
 
 	return nil
 }
 
-// equalTargets compares Target with each other.
-func equalTargets(t1, t2 Target) bool {
-	if t1.Role() != t2.Role() {
-		return false
-	}
+// toV2 writes Target to [v2acl.Target] messages of the NeoFS API protocol.
+func (t Target) toV2() []v2acl.Target {
+	var ms []v2acl.Target
 
-	keys1, keys2 := t1.BinaryKeys(), t2.BinaryKeys()
+	withKeys := len(t.keys) > 0
+	if withKeys {
+		ms = make([]v2acl.Target, len(t.roles)+1)
+	} else {
+		ms = make([]v2acl.Target, len(t.roles))
+	}
 
-	if len(keys1) != len(keys2) {
-		return false
+	for i := range t.roles {
+		ms[i].SetRole(v2acl.Role(t.roles[i]))
 	}
 
-	for i := 0; i < len(keys1); i++ {
-		if !bytes.Equal(keys1[i], keys2[i]) {
-			return false
-		}
+	if withKeys {
+		ms[len(ms)-1].SetKeys(t.keys)
```

Comment:
Right now, we have all the checks to be sure this row works appropriately. But one may change some of these checks, and in one moment `ms` can't be zero length => we have runtime panic.
Could we add a simple check `ms > 0`?